### PR TITLE
Prevent automatic batching updates in ThemeProvider to ensure client use right colorMode

### LINKED ISSUE #2731: Theming issue when using SSR
### Description

When the `colorMode` property of `<ThemeProvider />` is set to `auto` and SSR is used, it will still be displayed in light mode on browsers that prefer dark mode.

### Steps to reproduce

1. Set the `colorMode` property of `<ThemeProvider />` to `auto`.
2. Perform Next.js SSR / SSG rendering.
3. Open the page on the browser that prefers dark mode.
4. The displayed color scheme is `day`, not `night`.

### Temporary patch

Add a hook named `useAutoColorMode`.

```typescript
import { useEffect, useState } from 'react';

import useMediaQuery from '@/hooks/useMediaQuery';

enum ComponentColorMode {
  auto = 'auto',
  day = 'day',
  night = 'night',
}

const useAutoColorMode = () => {
  const [colorMode, setColorMode] = useState(ComponentColorMode.day);
  const darkModeEnabled = useMediaQuery('(prefers-color-scheme: dark)');

  useEffect(() => {
    setColorMode(
      darkModeEnabled ? ComponentColorMode.night : ComponentColorMode.day
    );
  }, [darkModeEnabled]);

  return colorMode;
};

export default useAutoColorMode;
```

Manually set `colorMode` in `App.tsx`.

```diff
const App = () => {
+  const colorMode = useAutoColorMode();

  // ...

  return (
-    <ThemeProvider colorMode="auto" preventSSRMismatch>
+    <ThemeProvider colorMode={colorMode} preventSSRMismatch>
```

### Version

v35.16.0

### Browser

Chrome

### PR DESCRIPTION
**Pull request by renbaoshuo**
Closes #2731.

Use `ReactDOM.flushSync` to prevent automatic batching of state updates since React 18, which causes error described in #2731.

### Screenshots

_None._

### Merge checklist

- [x] ~~Added/updated tests~~
- [x] ~~Added/updated documentation~~
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Comment by joshblack** [comment]
Thanks for taking the time to make Pull Request @renbaoshuo! 🙌 In your opinion, would there be a good test that we could add to make sure this doesn't happen again in the future?

File: src/ThemeProvider.tsx [hunk 1]
```diff
@@ -83,8 +84,13 @@ export const ThemeProvider: React.FC<React.PropsWithChildren<ThemeProviderProps>
         // if the resolved color mode passed on from the server is not the resolved color mode on client, change it!
         if (resolvedColorModePassthrough.current !== resolvedColorModeOnClient) {
           window.setTimeout(() => {
-            // override colorMode to whatever is resolved on the client to get a re-render
-            setColorMode(resolvedColorModeOnClient)
+            // use ReactDOM.flushSync to prevent automatic batching of state updates since React 18
+            // ref: https://github.com/reactwg/react-18/discussions/21
+            ReactDOM.flushSync(() => {
```
Comment (siddharthkp) [inline src/ThemeProvider.tsx:89]: Just checking, would this lead to an error with React 17?

Comment (renbaoshuo) [reply src/ThemeProvider.tsx:89]: https://github.com/facebook/react/pull/10225
https://github.com/facebook/react/issues/11527#issuecomment-360199710

It seems that `ReactDOM.flushSync(batch)` is already introduced before React 17.

**Comment by renbaoshuo** [comment]
> Thanks for taking the time to make Pull Request @renbaoshuo! 🙌 In your opinion, would there be a good test that we could add to make sure this doesn't happen again in the future?

Sorry, I have no idea.

**Review by joshblack — APPROVED** [review]

**Pull request merged**